kong/router/fields.lua

```diff
@@ -209,7 +219,67 @@ if is_http then
 
         return params.queries[field:sub(PREFIX_LEN + 1)]
       end
-    end
+
+    elseif field:sub(1, HTTP_SEGMENTS_PREFIX_LEN) == HTTP_SEGMENTS_PREFIX then
+      return function(params)
+        if not params.segments then
+          HTTP_SEGMENTS_REG_CTX.pos = 2 -- reset ctx, skip first '/'
+          params.segments = re_split(params.uri, "/", "jo", HTTP_SEGMENTS_REG_CTX)
+        end
+
+        local segments = params.segments
+
+        local name = field:sub(HTTP_SEGMENTS_PREFIX_LEN + 1)
+        local value = segments[name]
+
+        if value then
+          return value ~= ngx_null and value or nil
+        end
+
+        -- "/a/b/c" => 1="a", 2="b", 3="c"
+        -- http.path.segments.0 => params.segments[1 + 0] => a
+        -- http.path.segments.1_2 => b/c
+
+        local p = name:find("_", 1, true)
+
+        -- only one segment
+
+        if not p then
+          local pos = tonumber(name)
+
+          value = pos and segments[HTTP_SEGMENTS_OFFSET + pos] or nil
+          segments[name] = value or ngx_null
+
+          return value
+        end
+
+        -- (pos1, pos2) defines a segment range
+
+        local pos1 = tonumber(name:sub(1, p - 1))
+        local pos2 = tonumber(name:sub(p + 1))
+        local segs_count = #segments - HTTP_SEGMENTS_OFFSET
+
+        if not pos1 or not pos2 or
+           pos1 >= pos2 or pos1 > segs_count or pos2 > segs_count
+        then
+          segments[name] = ngx_null
+          return nil
+        end
+
+        local buf = buffer.new()
+
+        for p = pos1, pos2 - 1 do
+          buf:put(segments[HTTP_SEGMENTS_OFFSET + p]):put("/")
```

Comment:
``` suggestion
          buf:put(segments[HTTP_SEGMENTS_OFFSET + p], "/")
```

This one might be more readable.

Reply:
No, this form is a wrong call for string.buffer.

Reply:
> No, this form is a wrong call for string.buffer.

``` lua
local buf = require("string.buffer").new()
buf:put("Hello", " ", "World")
print(buf:get())
```

``` bash
$ resty test.lua
> Hello World
```

Did I miss something?

Reply:
done, thanks.